### PR TITLE
Fix amp-sidebar local navigation and history

### PR DESCRIPTION
**Pull request by mkhatib**
Debugging and experimenting with amp-sidebar bug #4184

Try the [demo for this](http://amp-mk-2.herokuapp.com/examples/viewer.html). 
- On 2nd tab
- Open Sidebar
- Click Section 1 link
  - Notice it would close the tab and scroll correctly to the section.
- Scroll back up and open sidebar again
- Click Section 2 link
  - Notice it would close the tab and scroll correctly to the section.
- Scroll back up and open sidebar again
- Click Section 1 link
  - Notice it would close the tab and scroll correctly to the section.
- Hit Back button 
  - Notice it should scroll to Section 2.
- Hit Back button Again
  - Notice it should scroll to Section 1.

### CODE REVIEW

File: examples/viewer.html [hunk 1]
```diff
@@ -381,10 +381,22 @@
...
 
-    Viewer.prototype.onPopState_ = function() {
-      // Even more trivial. Always assumes that only one step was popped in
-      // history.
-      this.stackIndex_--;
+    Viewer.prototype.onPopState_ = function(e) {
+      if (this.visibilityState_ != 'visible') {
+        return;
+      }
+      console.log('popping: ', e.state);
```
Comment (dvoytenko) [inline examples/viewer.html]: `log`

File: extensions/amp-sidebar/0.1/amp-sidebar.js [hunk 2]
```diff
@@ -103,6 +107,19 @@ export class AmpSidebar extends AMP.BaseElement {
     this.registerAction('toggle', this.toggle_.bind(this));
     this.registerAction('open', this.open_.bind(this));
     this.registerAction('close', this.close_.bind(this));
+
+    installGlobalClickListenerForDoc(this.getAmpDoc()).addBeforeHandler(e => {
```
Comment (dvoytenko) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: If we do something like this, I'd recommend we instead expose `onNavigate(href, target, isLocal)` method or something so that we don't have to look for all the same `A` and its parameters. Alternatively, I don't see an issue to just add a click listener directly on the `sidebar` itself. Why do we need to listen for global clicks anyway?

File: src/document-click.js [hunk 3]
```diff
@@ -37,7 +37,7 @@ const ORIGINAL_HREF_ATTRIBUTE = 'data-amp-orig-href';
  * @param {!./service/ampdoc-impl.AmpDoc} ampdoc
  */
 export function installGlobalClickListenerForDoc(ampdoc) {
-  fromClassForDoc(ampdoc, 'clickhandler', ClickHandler);
+  return fromClassForDoc(ampdoc, 'clickhandler', ClickHandler);
```
Comment (dvoytenko) [inline src/document-click.js]: Also add `@return`

File: src/document-click.js [hunk 4]
```diff
@@ -73,6 +73,9 @@ export class ClickHandler {
       this.boundHandle_ = this.handle_.bind(this);
       this.ampdoc.getRootNode().addEventListener('click', this.boundHandle_);
     }
+
+    /** @private {!Array<!function(!Event)>} */
+    this.beforeHandlers_ = [];
```
Comment (dvoytenko) [inline src/document-click.js]: We have observables for this.

File: extensions/amp-sidebar/0.1/amp-sidebar.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: Where is the listener removed? We don't need it working all the time, right? Only when the sidebar is open, I assume.

File: src/document-click.js [hunk 5]
```diff
@@ -91,9 +94,20 @@ export class ClickHandler {
    * @param {!Event} e
    */
   handle_(e) {
+    for (let i = 0; i < this.beforeHandlers_.length; i++) {
```
Comment (dvoytenko) [inline src/document-click.js]: ditto: observable

File: src/document-click.js [hunk 6]
```diff
@@ -286,10 +300,12 @@ export function onDocumentElementClick_(
   }
 
   if (tgtLoc.hash != curLoc.hash) {
+    timerFor(win).delay(() => {
```
Comment (dvoytenko) [inline src/document-click.js]: What is "wait" for?

File: extensions/amp-sidebar/0.1/amp-sidebar.js [hunk 7]
```diff
@@ -160,7 +177,9 @@ export class AmpSidebar extends AMP.BaseElement {
         }, ANIMATION_TIMEOUT);
       });
     });
-    this.getHistory_().push(this.close_.bind(this)).then(historyId => {
+    this.getHistory_().push(() => {
+      this.close_();
```
Comment (dvoytenko) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: I didn't understand the issue with event sequence. What's the exact event sequence in this case?

Comment (mkhatib) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: So without the doing the `delay` when pushing the new state in `document-submit` the sequence goes like this:

* Open Sidebar
* Click Section 1 Link
  * We Close Sidebar
  * Which also Pops History at 1
  * And then handle the actual click on the link and Push history to 1
  * And then Popping actually happen and history is popped to 0 removing the new push with it.

I am not entirely sure yet why does this happen. But this might be dependent on when the viewer is updating their stack index and communicate the change back with the amp doc. In `viewer.html` for example we do that inside the `popstate` event handler. This is were my theory about the event queue delaying the stack index update from happening right away after the close-pop.

 (I know the `viewer.html` implementation is incomplete but my guess is the Google.com viewer is doing something similar, I've also tested with the sandbox viewer and seen the same experience)

So to summarize this is what is being logged in sequence when the timeout is not in place.
```
[VIEWER] push history to  1
[VIEWER] pop history at  1
[VIEWER] push history to  1
[VIEWER] history popped to  0 # Notice our history push is gone now!
```

After adding the timeout this gives time for `popstate` handlers to update the stack indexes. These are the logs after timeout:
```
[VIEWER] push history to  1
[VIEWER] pop history at  1
[VIEWER] history popped to  0
[VIEWER] push history to  1 # Yay pushed state preserved!
```

Does this make any sense 😄 because I am not 80% share!

File: src/document-click.js (hunk 6 above)
Comment (mkhatib) [inline src/document-click.js]: Explained in the above discussion.

File: extensions/amp-sidebar/0.1/amp-sidebar.js (hunk 2 above)
Comment (mkhatib) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: Ah good point. Will update once other parts are figured out.

Comment (mkhatib) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: > Why do we need to listen for global clicks anyway?

This is basically to make sure we run before we actually handle the click (which is handled inside `document-click` handler), and the event is captured there and from testing capturing seems to be first-capture-first-serve and since `document-click` listener is added on document load it's always going to be the first handle to be called. 

In this case we want our handler inside sidebar to be called before it to allow us to pop before pushing.

**Comment by mkhatib** [comment]
cc/ @aghassemi and @camelburrito this is the PR we were discussing today.

File: extensions/amp-sidebar/0.1/amp-sidebar.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: I don't believe `document-click` is a capture. It'd seem to be illogical to have it as a capture click, since it's a classical bubbling processing.

File: extensions/amp-sidebar/0.1/amp-sidebar.js (hunk 7 above)
Comment (dvoytenko) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: So event processing loop can possibly interfere here. In this case, though, waiting for one event loop (e.g. `setTimeout(..., 0)`) should be enough. Can you try that in your experiments. We just need this info to make a better decision.

Generally, I'd say, we don't have here enough information to make the right calls. We need more research and understand the exact sequence of events much better. If we need more data from sandbox - we should find a way to get it. One important thing: our history supports queuing and returns promises for `push` and `pop`. In theory, these promises should be sufficient to address any risk conditions. For this purpose I'm a little skeptical that we need timers.

Comment (mkhatib) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: During myinitiall experiments I couldn't fix with a small timeout. I didn't investigate this enough bit wanted to see if you might have an idea why that might be happening. 

I'll do some more work on this and see if I can use the promises for this.

File: extensions/amp-sidebar/0.1/amp-sidebar.js (hunk 2 above)
Comment (mkhatib) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: Just checked you're right it's not capture. I could swear it was 😅 when I checked

File: extensions/amp-sidebar/0.1/amp-sidebar.js (hunk 7 above)
Comment (mkhatib) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: More of the "pop is done" - I don't think the same problem happens with pushing since there's no `pushstate` event.

**Comment by mkhatib** [comment]
@dvoytenko I cleaned up the PR PTAL 👀 - if this looks good I'll add tests. I'll also test with the sandbox viewer.

**Comment by dvoytenko** [comment]
@mkhatib Is this ready to be reviewed again?

**Comment by mkhatib** [comment]
@dvoytenko yes, I'll work on tests

File: extensions/amp-sidebar/0.1/amp-sidebar.js [hunk 8]
```diff
@@ -107,14 +110,25 @@ export class AmpSidebar extends AMP.BaseElement {
...
+    screenReaderCloseButton.addEventListener('click', () => this.close_());
     this.element.appendChild(screenReaderCloseButton);
 
     this.registerAction('toggle', this.toggle_.bind(this));
     this.registerAction('open', this.open_.bind(this));
     this.registerAction('close', this.close_.bind(this));
+
+    this.element.addEventListener('click', e => {
+      const target = closestByTag(dev().assertElement(e.target), 'A');
+      if (!target) {
```
Comment (jridgewell) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: Check that `target` has a `href`, too. Also, this code's [duplicated](https://github.com/ampproject/amphtml/blob/553492cf2c54e9c974ddea12b3c524a03b68de0b/src/document-click.js#L116-L124).

File: extensions/amp-sidebar/0.1/amp-sidebar.js [hunk 9]
```diff
@@ -107,14 +110,25 @@ export class AmpSidebar extends AMP.BaseElement {
     screenReaderCloseButton.classList.add('-amp-screen-reader');
     // This is for screen-readers only, should not get a tab stop.
     screenReaderCloseButton.tabIndex = -1;
-    screenReaderCloseButton.addEventListener('click', () => {
-      this.close_();
-    });
+    screenReaderCloseButton.addEventListener('click', () => this.close_());
```
Comment (jridgewell) [inline extensions/amp-sidebar/0.1/amp-sidebar.js]: It seems like we should create a `boundClose`, since it's used so frequently.

File: src/service/history-impl.js [hunk 10]
```diff
@@ -601,14 +611,25 @@ export class HistoryBindingVirtual_ {
...
+    // Waiting for popping confirmation to happen otherwise a race condition
+    // between popping and pushing might happen causing history to be messed
+    // up. This will only wait 100ms to avoid getting stuck if confirmation
+    // never arrived.
+    return Promise.race([this.timer_.promise(100), this.awaitPoppingPromise_])
+        .then(() => {
+          // Current implementation doesn't wait for response from viewer.
+          this.updateStackIndex_(this.stackIndex_ + 1);
+          this.viewer_.postPushHistory(this.stackIndex_);
+          return Promise.resolve(this.stackIndex_);
```
Comment (jridgewell) [inline src/service/history-impl.js]: Just return `#stackIndex`.

File: src/service/history-impl.js [hunk 11]
```diff
@@ -573,8 +573,12 @@ export class HistoryBindingVirtual_ {
 
   /**
    * @param {!./viewer-impl.Viewer} viewer
+   * @param {!Window} win
    */
-  constructor(viewer) {
+  constructor(viewer, win) {
+    /** @private @const {!../service/timer-impl.Timer} */
+    this.timer_ = timerFor(win);
```
Comment (dvoytenko) [inline src/service/history-impl.js]: You can just do `viewer.ampdoc.win`. No need to pass `win` as an arg here.

File: src/service/history-impl.js [hunk 12]
```diff
@@ -601,14 +611,25 @@ export class HistoryBindingVirtual_ {
...
   push() {
-    // Current implementation doesn't wait for response from viewer.
-    this.updateStackIndex_(this.stackIndex_ + 1);
-    this.viewer_.postPushHistory(this.stackIndex_);
-    return Promise.resolve(this.stackIndex_);
+    // Waiting for popping confirmation to happen otherwise a race condition
+    // between popping and pushing might happen causing history to be messed
+    // up. This will only wait 100ms to avoid getting stuck if confirmation
+    // never arrived.
+    return Promise.race([this.timer_.promise(100), this.awaitPoppingPromise_])
```
Comment (dvoytenko) [inline src/service/history-impl.js]: Why `awaitPoppingPromise_` necessary? Isn't queuing feature already provided by the parent `History` class? It'd seem all we need to do is return a promise from viewer's `postPushHistory` call, which may already be the case.

**Comment by mkhatib** [comment]
Closing this one for #5961
